libs/wingc/src/type_check/symbol_env.rs

```diff
@@ -179,71 +211,69 @@ impl SymbolEnv {
 		self: reference([Self]),
 		symbol: &Symbol,
 		not_after_stmt_idx: Option<usize>,
-		depth: usize,
 	) -> Option<reference([SymbolKind])> {
-		self
-			.lookup_ext(symbol, not_after_stmt_idx, depth)
-			.ok()
-			.map(|(kind, _)| kind)
+		self.lookup_ext(symbol, not_after_stmt_idx).ok().map(|(kind, _)| kind)
 	}
 
 	#[allow(clippy::needless_arbitrary_self_type)]
 	#[duplicate_item(
-		lookup_ext LookupResult map_get reference(type) ref_annotation(ident);
-		[lookup_ext] [LookupResult] [get] [& type] [ref ident];
-		[lookup_ext_mut] [LookupResultMut] [get_mut] [&mut type] [ref mut ident];
+		lookup_ext LookupResult map_get reference(type) ref_annotation(ident) get_ref SymbolLookupInfo;
+		[lookup_ext] [LookupResult] [get] [& type] [ref ident] [self.get_ref()] [SymbolLookupInfo];
+		[lookup_ext_mut] [LookupResultMut] [get_mut] [&mut type] [ref mut ident] [()] [SymbolLookupInfoMut];
 	)]
 	/// Lookup a symbol in the environment, returning a `LookupResult`. Note that the symbol name
 	/// cannot be a nested symbol (e.g. `foo.bar`), use `lookup_nested` for that.
 	/// TODO: perhaps make this private and switch to the nested version in all external calls
-	pub fn lookup_ext(
-		self: reference([Self]),
-		symbol: &Symbol,
-		not_after_stmt_idx: Option<usize>,
-		depth: usize,
-	) -> LookupResult {
+	pub fn lookup_ext(self: reference([Self]), symbol: &Symbol, not_after_stmt_idx: Option<usize>) -> LookupResult {
+		let mut result = LookupResult::NotFound(symbol.clone());
+
 		if let Some((definition_idx, kind)) = self.symbol_map.map_get(&symbol.name) {
-			if let Some(not_after_stmt_idx) = not_after_stmt_idx {
-				if let StatementIdx::Index(definition_idx) = definition_idx {
-					if *definition_idx > not_after_stmt_idx {
-						return LookupResult::DefinedLater;
-					}
-				}
+			// if found the symbol and it is defined before the statement index (or statement index is
+			// unspecified, which is likely not something we want to support), we found it
+			let lookup_index = not_after_stmt_idx.unwrap_or(usize::MAX);
+			let definition_index = match definition_idx {
+				StatementIdx::Top => 0,
+				StatementIdx::Index(idx) => *idx,
+			};
+
+			// found it!
+			if lookup_index >= definition_index {
+				return LookupResult::Found(
+					kind,
+					SymbolLookupInfo {
+						phase: self.phase,
+						init: self.is_init,
+						env: get_ref,
+					},
+				);
 			}
-			LookupResult::Found(
-				kind,
-				SymbolLookupInfo {
-					phase: self.phase,
-					init: self.is_init,
-					depth: depth,
-				},
-			)
-		} else if let Some(ref_annotation([parent_env])) = self.parent {
-			parent_env.lookup_ext(symbol, not_after_stmt_idx.map(|_| self.statement_idx), depth + 1)
+
+			// its defined later in the current scope, but we still need to check if it might be captured
+			// from the parent.
```

Comment:
Not sure this comment is understandable in this context.
Also I'm not sure what the changes you did in this function do, seems like the same logic as before. Noting this in case I missed something.

Reply:
mmm.. actually, I think the previous implementation was the correct one. It returned "defined later" even if the symbol could be resolved from a parent environment. Originally I thought that's something we wanted to support, but we realized this cannot be implemented in JavaScript so basically I think we can revert this and remove the check I added in the capture scanner.

Reply:
Yap, indeed. Reverted the previous behavior (with slightly more readable code if that's okay).